### ISSUE TITLE
Slow Page Loading

### ISSUE THREAD

**Issue by lukeed** [pyrocms/pyrocms#3525]
Hopping around the admin area takes 3 to 5 seconds between each page, not sure why

* Homestead v2.0.7, default settings + higher fastcgi_read_timeout threshold ( #3524  ) 
* Vagrant v1.6.3
* Using MySql connection

Everything works; just slowly :crying_cat_face:

**Comment by RyanThompson** [comment]
Hmm - I know there are improvements to be made but that's a little outlandish. Let's see if we can dig into it a bit:

1.) Is your debug mode enabled? This will cause things to run slower and should ONLY be ran when debugging (still I get about .8 / 1 second load with debug = true)

2.) Take a look at your debug bar when you enable it - what are the times? Queries?

3.) In the lower right of the screen you'll have memory / time footprint. What do they say now?

**Comment by lukeed** [comment]
Sorry, should have included these initially :zzz: 

1) Debug Off: 4.6s. Debug On: 5.3s. Both on dashboard

2)
Timeline (lot of nothingness):
![screen shot 2015-08-18 at 12 10 28 pm](https://cloud.githubusercontent.com/assets/5855893/9340004/28429432-45a2-11e5-90e5-03d03ca50daf.png)

Queries (5, normal):
![screen shot 2015-08-18 at 12 11 27 pm](https://cloud.githubusercontent.com/assets/5855893/9340022/4a97f9fa-45a2-11e5-9b78-494de65c8f64.png)

3) Debug Off: (4.2s - 4.8s)/17.3MB. Debug On: (5.2s - 5.9s)/18.75MB

**Comment by RyanThompson** [comment]
That’s quite the gap between your booting and application..

Can you put this:

    dd(number_format(microtime(true) - $_SERVER['REQUEST_TIME_FLOAT'], 2) . ' s’);

Here:

https://github.com/anomalylabs/streams-platform/blob/1.0/develop/src/StreamsServiceProvider.php#L55

And let perhaps before the loading of addons too and let me know what you find for times.



Ryan Thompson
President / CEO
AnomalyLabs, Inc.
www.anomaly.is<http://www.anomaly.is>

t (877) 705-0806
m (309) 752-8581



On Aug 18, 2015, at 2:13 PM, Luke Edwards <notifications@github.com<mailto:notifications@github.com>> wrote:


Sorry, should have included these initially [:zzz:]

1) Debug Off: 4.6s. Debug On: 5.3s. Both on dashboard

2)
Timeline (lot of nothingness):
[screen shot 2015-08-18 at 12 10 28 pm]<https://cloud.githubusercontent.com/assets/5855893/9340004/28429432-45a2-11e5-90e5-03d03ca50daf.png>

Queries (5, normal):
[screen shot 2015-08-18 at 12 11 27 pm]<https://cloud.githubusercontent.com/assets/5855893/9340022/4a97f9fa-45a2-11e5-9b78-494de65c8f64.png>

3) Debug Off: (4.2s - 4.8s)/17.3MB. Debug On: (5.2s - 5.9s)/18.75MB

**Comment by lukeed** [comment]
Sure. Probably want to hide your phone number here

**Comment by RyanThompson** [comment]
haha it's everywhere..

**Comment by lukeed** [comment]
Bold.

Where | Time (s) | Comment
--------- | ----------- | -------------
Before dispatchers | 0.05 - 0.75 | slight disparity
Before `$this->app->booted()` function | 0.16 - 1.08 | disparity
Inside `$this->app->booted()` function | 3.07 | constant
After `$this->app->booted()` function | 0.15 - 1.19 | disparity

**Comment by RyanThompson** [comment]
Is that **inside boot** one after registering addons or before? I'd like to know both if you can please.

Does homestead use NFS by default?

**Comment by lukeed** [comment]
Not using NFS.

Before registering: 0.12s to 1.2s
After registering: 2.9s to 3.15s

**Comment by RyanThompson** [comment]
NFS will improve this drastically

Sent from my iPhone

On Aug 18, 2015, at 2:51 PM, Luke Edwards <notifications@github.com<mailto:notifications@github.com>> wrote:


Not using NFS.

Before registering: 0.12s to 1.2s
After registering: 2.9s to 3.15s

-

**Comment by lukeed** [comment]
You'll have to forgive me for this as I don't fully understand how or why NFS offers the performance boost that it does, but...

I know that NFS offers a performance boost across the board for mostly all apps. Ergo, I try to build the most performant app I can before I start turning on the steroids. That way, I end up with the fastest possible native/natural app, so that I am not solely relying on boosters.

(Plus, I vaguely recall that Vagrant NFS won't work on Windows machines?)

Anyway, I'm sure that will solve my current issue, but any thoughts on what can be done to fix this without NFS?

**Comment by RyanThompson** [comment]
Hey,

First off, I totally agree. And I am the first one to point out a wee bit of a performance issue. I am assembling a small team to look into it and beat the issue until it's dead. I know exactly WHERE the issue is but I don't entirely know why (which will lead to a speedy fix).

That said though, your times, for whatever reason, are out of the norm.

First thing is first, I am almost CERTAIN that your issue in particular is coming down to disk IO bottlenecking. Laravel, it's dependencies and the Streams Platform in PyroCMS are all OOP and PSR4 with lots of small classes. Something is causing your IO to shit its pants I believe.

On a typical setup, IO only accounts for like .02 seconds but I've seen some issues on virtual machines that blow it out of the water. **NFS will tell your machine NOT to share disk resources so file lookups are exponentially faster.**

As far as the performance issue - all I can say is that it WILL be fixed before the stable release. It quite literally has to be. On top of that there are already some features I've implemented to help curb it in the meanwhile lol. PyroCMS.com is, after all, on V3 and a VERY modest server.

Windows.. That's a huge red flag for me in general. Some of the other well versed community members have described problems with PyroCMS, that are 100% out of PyroCMSs control (like long zip paths when composer is doing it's thing). And unfortunately I am not a windows guy so it's a little confusing for me sometimes.

So long story short, I don't wanna cop out and boost the shit out of it lol. But NFS is a typical disk mode for a server and I know it speeds things up. ALSO, rest assured the performance will be handled come hell or high water.

Also - the addon registering is the slow part.. In particular event listeners and certain binding resolution so I am working on identifying in full the issue and rewriting the addon PSR / Laravel integration sequence. 

Hopefully this answers your question and sheds some light on things ^_^

PS if you are interested in helping out I have a dedicated Slack group to this. Just email me (or text me - you have my mobile number :-P) your username and Ill invite you to the group.

Cheers!

Ryan

**Comment by lukeed** [comment]
Dang, great answer -- thanks! Informative & lol-worthy even.Too bad it's just an issue reply!

Fully agreed on all accounts, and I haven't gone near a Windows machine in years. I'm surely with you there... 

Im not sure about the bottlenecking either, though, as I said, I'm just running with default Homestead settings. So that means:
```
memory: 2048
cpus: 1
```

I also have nearly 10 sites linked up on the machine. They're all fairly small projects, but perhaps, collectively, they are still using enough disk resources to cause a hefty function (like add-on registering) to make Homestead "shit its pants", lol

I'd love to help out! Bad news is that I can't for the next couple weeks (#deadlines), but I'll surely email you when I can be of service. That kind of group is something I wanna be a part of.

Best ~

**Comment by RyanThompson** [comment]
Ill be hearing from you then! Until then I will ping you here with updates (I've already spent some time on it last night).

**Comment by ryancwalsh** [comment]
I may have resolved this. 

I'm running Laravel Homestead on Windows 10, and my pages are way too slow, which is why I bailed on PyroCMS the first time I took a peek at it. (3.14 seconds for /admin/dashboard/view/welcome)

https://pyrocms.com/forum/channels/general/slow-page-loads says to set `type: nfs`.

So I installed the "winnfsd" plugin from https://github.com/winnfsd/vagrant-winnfsd/tree/f703e8996bbb01312755b554689cb910b92426b9 and modified my Homestead.yaml's folder section (*not* the Sites section) to have:

      type: "nfs"
      mount_options: ['nolock,vers=3,udp,noatime']

Then I ran `homestead.bat reload --provision`. 

Now my `/admin/dashboard/view/welcome` loaded in 0.37s.

**Comment by ultrono** [comment]
Have an old Windows laptop running Windows 10. I was experiencing 8-12 second load times for the admin dashboard screen. After enabling NFS via "winnfsd" that dropped to < 2.5 seconds 👍 

I have other Laravel sites on the same Windows laptop and they are fine. I assume Pyro (v3.5) is loading a lot more files.

When I switch to Ubuntu on the same laptop the load times are < 1.5 seconds :)

Moral of the story - NFS is basically essential for Windows development (if peeps still put themselves through that nowadays?) as 8-12 seconds load times in development essentially makes Pyro an nonviable choice (which is a shame, because as CMS's go, it's up there with some of the best IMO).
